### DISCUSSION TITLE
shopify hmac computed not the same as received one

### DISCUSSION THREAD

**Discussion by Braingearceo** [fastapi/fastapi#11693] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import hmac
import hashlib
import base64
from fastapi import FastAPI, Request, HTTPException
from fastapi.responses import Response
import os

load_dotenv()

app = FastAPI()

SHOPIFY_SECRET = os.getenv('SHOPIFY_SECRET')

def verify_webhook(raw_data, hmac_header, secret):
    digest = hmac.new(secret.encode('utf-8'), raw_data, digestmod=hashlib.sha256).digest()
    computed_hmac = base64.b64encode(digest).decode()

    print(f"SECRET: {secret}")
    print(f"RAW DATA: {raw_data}")
    print(f"RAW DATA (decoded): {raw_data.decode('utf-8')}")
    print(f"DIGEST (binary): {digest}")
    print(f"DIGEST (hex): {digest.hex()}")
    print(f"COMPUTED HMAC: {computed_hmac}")
    print(f"RECEIVED HMAC: {hmac_header}")

    return hmac.compare_digest(computed_hmac, hmac_header)

@app.post('/webhook/customer_data_request')
async def customer_data_request_webhook(request: Request):
    try:
        raw_data = await request.body()
        headers = dict(request.headers)
        print('HEADERS:', headers)
        print('RAW DATA:', raw_data)
        hmac_header = headers.get('X-Shopify-Hmac-SHA256')
        if not hmac_header:
            raise HTTPException(status_code=400, detail="Missing HMAC header")
        verified = verify_webhook(raw_data, hmac_header, SHOPIFY_SECRET)
        if not verified:
            print("Computed HMAC does not match the provided HMAC.")
            raise HTTPException(status_code=401, detail="HMAC verification failed")

        print("HMAC verification successful.")
        return Response(status_code=200)

    except Exception as e:
        print("Error processing customer data request webhook:")
        print(e)
        return Response(status_code=500, content=str(e))
```

### Description

I’m building a Shopify public app and I’m having trouble with the HMAC verification process. Despite using the Shopify Secret from the Dashboard of my app at partners.shopify.com, I can’t get the computed HMAC and the received HMAC to match. This always results in a failed verification.

I’ve tried various combinations, but nothing seems to work—they never match. There might be something I’m missing. I hope someone here has a solution.

Currently, they provide the following Flask example (that do work). 

```
# The following example uses Python and the Flask framework to verify a webhook request:

from flask import Flask, request, abort
import hmac
import hashlib
import base64

app = Flask(__name__)

# The Shopify app's client secret, viewable from the Partner Dashboard. In a production environment, set the client secret as an environment variable to prevent exposing it in code.
CLIENT_SECRET = 'my_client_secret'

def verify_webhook(data, hmac_header):
    digest = hmac.new(CLIENT_SECRET.encode('utf-8'), data, digestmod=hashlib.sha256).digest()
    computed_hmac = base64.b64encode(digest)

    return hmac.compare_digest(computed_hmac, hmac_header.encode('utf-8'))

@app.route('/webhook', methods=['POST'])
def handle_webhook():
    data = request.get_data()
    verified = verify_webhook(data, request.headers.get('X-Shopify-Hmac-SHA256'))

    if not verified:
        abort(401)

    # Process webhook payload
    # ...

    return ('', 200)

```

So I can't figure out what I'am doing wrong...

### Operating System

macOS

### FastAPI Version

0.110.1

### Pydantic Version

2.7.3

### Python Version

3.11.9

**Comment by jgould22** [comment]
First step here would be to start systematically removing things to remove possible sources of error.

HMAC, like any cryptographic algo is really picky and likes to use bytes.

I would start by hard coding your key instead of setting it in the env since that is one of the most obvious differences between your code and your flask sample.

**Reply by Braingearceo** [reply]
Hey, thank you for your suggestion! I already tried to hardcode the key but it still compute the wrong hmac. 

Is there a difference between Flask request.get_data(cache=True) and request.get_data() in FastAPI ? Like await request.body() return the equivalent of request.get_data(cache=True) isn't it? 

btw, here are the printed logs :

RAW DATA: b'{"shop_id":69161943062,"shop_domain":"app-security-02.myshopify.com"}'
HEADERS: {'host': '******************', 'user-agent': 'Ruby', 'content-length': '69', 'accept': '*/*', 'accept-encoding': 'gzip', 'cdn-loop': 'cloudflare; subreqs=1', 'cf-connecting-ip': '34.139.67.206', 'cf-ew-via': '15', 'cf-ipcountry': 'US', 'cf-ray': '88ff3142c1924572-ATL', 'cf-visitor': '{"scheme":"https"}', 'cf-worker': 'onrender.com', 'content-type': 'application/json', 'render-proxy-ttl': '4', 'rndr-id': '60f851c9-2642-4d3c', 'traceparent': '00-6754ba8eb108e5859468b00673dd757f-7eafd478f3e913c3-01', 'true-client-ip': '34.139.67.206', 'x-cloud-trace-context': '6754ba8eb108e5859468b00673dd757f/9128748585656587203;o=1', 'x-forwarded-for': '34.139.67.206, 10.216.42.201', 'x-forwarded-proto': 'https', 'x-request-start': '1717748106965652', 'x-shopify-api-version': '2022-10', 'x-shopify-hmac-sha256': 'g+z4klhDapK3zSJjLj9tL8iptRkwqLVPGWk2dhopmf4=', 'x-shopify-shop-domain': 'app-security-02.myshopify.com', 'x-shopify-test': 'true', 'x-shopify-topic': 'shop/redact', 'x-shopify-trace-context': '6754ba8eb108e5859468b00673dd757f/9128748585656587203;o=1', 'x-shopify-trace-hint': 'true', 'x-shopify-triggered-at': '2024-06-07T08:15:06.611236172Z', 'x-shopify-webhook-id': '20467b62-e0b6-4bcf-879e-dbd5f0b61910'}
HMAC HEADER: g+z4klhDapK3zSJjLj9tL8iptRkwqLVPGWk2dhopmf4=
SECRET: b'***********************'
DATA (raw): b'{"shop_id":69161943062,"shop_domain":"app-security-02.myshopify.com"}'
DIGEST (binary): b'\xe5\x96K\rVp\xa3\xae\xc6\xb5\xc6a[*\x06\x12\xb1\xa9\xa0s5sE\n\x84p\xcb\xc8\x052\xeax'
DIGEST (hex): e5964b0d5670a3aec6b5c6615b2a0612b1a9a0733573450a8470cbc80532ea78
COMPUTED HMAC (base64): b'5ZZLDVZwo67GtcZhWyoGErGpoHM1c0UKhHDLyAUy6ng='
RECEIVED HMAC (header): g+z4klhDapK3zSJjLj9tL8iptRkwqLVPGWk2dhopmf4=
RECEIVED HMAC ENC: g+z4klhDapK3zSJjLj9tL8iptRkwqLVPGWk2dhopmf4=

**Reply by jgould22** [reply]
This is the body function in fastapi/starlette 
https://github.com/encode/starlette/blob/master/starlette/requests.py#L241

Flask if I recall uses werkzeug 
https://github.com/pallets/werkzeug/blob/main/src/werkzeug/wrappers/request.py#L388

Without digging too deeply it does look like they are similar 

ah wait

remove the decode from this call 
```
    computed_hmac = base64.b64encode(digest).decode()
```

**Reply by Braingearceo** [reply]
It's looks like the problem lies within what should be used to compute the hmac, since we can't really find a difference between request.data and request.get_data() equivalent in fastAPI, the raw body is correctly returned by await request.body().
Thank you for your response!

**Reply by Braingearceo** [reply]
from flask import Flask, request, abort
import hmac
import hashlib
import base64

app = Flask(__name__)

# The Shopify app's client secret. Set this as an environment variable in production.
CLIENT_SECRET = 'SECRET'

def verify_webhook(data, hmac_header):
    print(hmac_header)
    digest = hmac.new(CLIENT_SECRET.encode('utf-8'), data, digestmod=hashlib.sha256).digest()
    computed_hmac = base64.b64encode(digest)
    return hmac.compare_digest(computed_hmac, hmac_header.encode('utf-8'))

@app.route('/webhook/customer/data_request', methods=['POST'])
def customer_data_request_webhook():
    data = request.data
    print(data)
    verified = verify_webhook(data, request.headers.get('X-Shopify-Hmac-SHA256'))
    if not verified:
        abort(401)
    return '', 200

@app.route('/webhook/customer/data_erasure', methods=['POST'])
def customer_data_erasure_webhook():
    data = request.data
    print(data)
    verified = verify_webhook(data, request.headers.get('X-Shopify-Hmac-SHA256'))
    if not verified:
        abort(401)
    return '', 200

@app.route('/webhook/shop/data_erasure', methods=['POST'])
def shop_data_erasure_webhook():
    data = request.data
    print(data)
    verified = verify_webhook(data, request.headers.get('X-Shopify-Hmac-SHA256'))
    if not verified:
        abort(401)
    return '', 200

if __name__ == '__main__':
    app.run(port=8000)
    
If I use this exact code, it does work, I just tried. So it's surely is with that await request.body() bytes

**Reply by jgould22** [reply]
The above code is not calling decode on your computed hmac

You example in this post is 
```python
def verify_webhook(raw_data, hmac_header, secret):
    digest = hmac.new(secret.encode('utf-8'), raw_data, digestmod=hashlib.sha256).digest()
    computed_hmac = base64.b64encode(digest).decode() # <<-- REMOVE THIS '.decode()'
```

**Comment by YuriiMotov** [comment]
Looks like the author is not interested in this anymore. 
Let's close this for now to keep the discussions clean.
If you’ve found a solution, please share it here.
If you still need help, feel free to let us know.
